src/pg4.rs

```diff
@@ -0,0 +1,1029 @@
+/// use spin_sdk::pg4::{Connection, Decode};
+///
+/// # fn main() -> anyhow::Result<()> {
+/// let db = Connection::open("host=localhost user=postgres password=my_password dbname=mydb")?;
+///
+/// let query_result = db.query("SELECT COUNT(*) FROM users", &[])?;
+///
+/// assert_eq!(1, query_result.columns.len());
+/// assert_eq!("count", query_result.columns[0].name);
+/// assert_eq!(1, query_result.rows.len());
+///
+/// let count = i64::decode(&query_result.rows[0][0])?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Delete rows from a PostgreSQL table. This uses [Connection::execute()]
+/// instead of the `query` method.
+///
+/// ```no_run
+/// use spin_sdk::pg4::Connection;
+///
+/// # fn main() -> anyhow::Result<()> {
+/// let db = Connection::open("host=localhost user=postgres password=my_password dbname=mydb")?;
+///
+/// let rows_affected = db.execute(
+///     "DELETE FROM users WHERE name = $1",
+///     &["Baldrick".to_owned().into()]
+/// )?;
+/// # Ok(())
+/// # }
+/// ```
+#[doc(inline)]
+pub use super::wit::pg4::Connection;
+
+/// The result of a database query.
+///
+/// # Examples
+///
+/// Load a set of rows from a local PostgreSQL database, and iterate over them
+/// selecting one field from each. The columns collection allows you to find
+/// column indexes for column names; you can bypass this lookup if you name
+/// specific columns in the query.
+///
+/// ```no_run
+/// use spin_sdk::pg4::{Connection, Decode};
+///
+/// # fn main() -> anyhow::Result<()> {
+/// # let min_age = 0;
+/// let db = Connection::open("host=localhost user=postgres password=my_password dbname=mydb")?;
+///
+/// let query_result = db.query(
+///     "SELECT * FROM users WHERE age >= $1",
+///     &[min_age.into()]
+/// )?;
+///
+/// let name_index = query_result.columns.iter().position(|c| c.name == "name").unwrap();
+///
+/// for row in &query_result.rows {
+///     let name = String::decode(&row[name_index])?;
+///     println!("Found user {name}");
+/// }
+/// # Ok(())
+/// # }
+/// ```
+pub use super::wit::pg4::RowSet;
+
+#[doc(inline)]
+pub use super::wit::pg4::{Error as PgError, *};
+
+/// The PostgreSQL INTERVAL data type.
+pub use crate::pg4::Interval;
+
+use chrono::{Datelike, Timelike};
+
+/// A Postgres error
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Failed to deserialize [`DbValue`]
+    #[error("error value decoding: {0}")]
+    Decode(String),
+    /// Postgres query failed with an error
+    #[error(transparent)]
+    PgError(#[from] PgError),
+}
+
+/// A type that can be decoded from the database.
+pub trait Decode: Sized {
+    /// Decode a new value of this type using a [`DbValue`].
+    fn decode(value: &DbValue) -> Result<Self, Error>;
+}
+
+impl<T> Decode for Option<T>
+where
+    T: Decode,
+{
+    fn decode(value: &DbValue) -> Result<Self, Error> {
+        match value {
+            DbValue::DbNull => Ok(None),
+            v => Ok(Some(T::decode(v)?)),
+        }
+    }
+}
+
+impl Decode for bool {
```

Comment:
A macro to generate the impls for all of these primitive types might make things easier to read 🤷

Reply:
How strongly do you feel about doing that as part of this PR?  The code here follows `pg/pg3`: I like the idea of a macro but I assume we'd prefer to apply it across all versions (or all versions that we retain).

Reply:
I don't feel strongly about this. Just a thought.